File: chapter-05-permissions/csharp/PermissionOperations.cs
/*
 * PermissionOperations.cs
 * Chapter 05: Permissions
 *
 * SharePoint Permission Operations Example
 * Demonstrates managing sharing links and permissions
 */

using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using LaExperiencia.SharePoint.Common;
using Microsoft.Graph;
using Microsoft.Graph.Models;
using Microsoft.Graph.Drives.Item.Items.Item.CreateLink;
using Microsoft.Graph.Drives.Item.Items.Item.Invite;

namespace LaExperiencia.SharePoint.Chapter05.Permissions
{
    /// <summary>
    /// Example class demonstrating SharePoint permission operations.
    /// The Graph client is injected via constructor (DI) from the common module.
    /// </summary>
    public class PermissionOperations
    {
        private readonly GraphServiceClient _graphClient;

        /// <summary>Creates an instance with an injected Graph client (DI).</summary>
        public PermissionOperations(GraphServiceClient graphClient)
        {
            _graphClient = graphClient ?? throw new ArgumentNullException(nameof(graphClient));
        }

        /// <summary>
        /// Creates an anonymous sharing link for a file or folder
        /// </summary>
        /// <param name="siteId">The site ID</param>
        /// <param name="driveId">The drive ID</param>
        /// <param name="itemId">The item ID</param>
        /// <param name="linkType">The link type (view or edit)</param>
        /// <returns>The created permission with sharing link</returns>
        public async Task<Permission?> CreateAnonymousLinkAsync(
            string siteId,
            string driveId,
            string itemId,
            string linkType = "view")
        {
            try
            {
                Console.WriteLine($"Creating anonymous sharing link for item: {itemId}");

                var requestBody = new CreateLinkPostRequestBody
                {
                    Type = linkType,
                    Scope = "anonymous"
                };

                var permission = await _graphClient
                    .Drives[driveId]
                    .Items[itemId]
                    .CreateLink
                    .PostAsync(requestBody);

                Console.WriteLine("Sharing link created successfully:");
                Console.WriteLine($"  Link: {permission?.Link?.WebUrl}");
                Console.WriteLine($"  Type: {permission?.Link?.Type}");
                Console.WriteLine($"  Scope: {permission?.Link?.Scope}");

                return permission;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error creating sharing link: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Creates an organization sharing link (for users within the organization)
        /// </summary>
        /// <param name="siteId">The site ID</param>
        /// <param name="driveId">The drive ID</param>
        /// <param name="itemId">The item ID</param>
        /// <param name="linkType">The link type (view or edit)</param>
        /// <returns>The created permission with sharing link</returns>
        public async Task<Permission?> CreateOrganizationLinkAsync(
            string siteId,
            string driveId,
            string itemId,
            string linkType = "view")
        {
            try
            {
                Console.WriteLine($"Creating organization sharing link for item: {itemId}");

                var requestBody = new CreateLinkPostRequestBody
                {
                    Type = linkType,
                    Scope = "organization"
                };

                var permission = await _graphClient
                    .Drives[driveId]
                    .Items[itemId]
                    .CreateLink
                    .PostAsync(requestBody);

                Console.WriteLine("Organization sharing link created successfully:");
                Console.WriteLine($"  Link: {permission?.Link?.WebUrl}");
                Console.WriteLine($"  Type: {permission?.Link?.Type}");
                Console.WriteLine($"  Scope: {permission?.Link?.Scope}");

                return permission;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error creating organization link: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Grants access to a specific user
        /// </summary>
        /// <param name="siteId">The site ID</param>
        /// <param name="driveId">The drive ID</param>
        /// <param name="itemId">The item ID</param>
        /// <param name="userEmail">The email address of the user</param>
        /// <param name="role">The role (write or read)</param>
        /// <returns>The created permission</returns>
        public async Task<Permission?> GrantAccessToUserAsync(
            string siteId,
            string driveId,
            string itemId,
            string userEmail,
            string role = "write")
        {
            try
            {
                Console.WriteLine($"Granting {role} access to {userEmail} for item: {itemId}");

                var requestBody = new InvitePostRequestBody
                {
                    Recipients = new List<DriveRecipient>
                    {
                        new DriveRecipient
                        {
                            Email = userEmail
                        }
                    },
                    Roles = new List<string> { role },
                    SendInvitation = true,
                    Message = "You have been granted access to this document."
                };

                var permissions = await _graphClient
                    .Drives[driveId]
                    .Items[itemId]
                    .Invite
                    .PostAsInvitePostResponseAsync(requestBody);

                Console.WriteLine("Access granted successfully:");
                if (permissions?.Value is { Count: > 0 } perms)
                {
                    var permission = perms[0];
                    Console.WriteLine($"  Permission ID: {permission.Id}");
                    Console.WriteLine($"  Roles: {string.Join(", ", permission.Roles ?? new List<string>())}");
                }

                return permissions?.Value?[0];
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error granting access: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Lists all permissions for an item
        /// </summary>
        /// <param name="siteId">The site ID</param>
        /// <param name="driveId">The drive ID</param>
        /// <param name="itemId">The item ID</param>
        public async Task ListPermissionsAsync(string siteId, string driveId, string itemId)
        {
            try
            {
                Console.WriteLine($"Listing permissions for item: {itemId}");

                var permissions = await _graphClient
                    .Drives[driveId]
                    .Items[itemId]
                    .Permissions
                    .GetAsync();

                Console.WriteLine($"Found {permissions?.Value?.Count ?? 0} permissions:");
                Console.WriteLine(new string('-', 80));

                foreach (var permission in permissions?.Value ?? new List<Permission>())
                {
                    Console.WriteLine($"Permission ID: {permission.Id}");
                    Console.WriteLine($"  Roles: {string.Join(", ", permission.Roles ?? new List<string>())}");

                    if (permission.Link != null)
                    {
                        Console.WriteLine($"  Link Type: {permission.Link.Type}");
                        Console.WriteLine($"  Link Scope: {permission.Link.Scope}");
                        Console.WriteLine($"  Web URL: {permission.Link.WebUrl}");
                    }

                    if (permission.GrantedTo?.User != null)
                    {
                        Console.WriteLine($"  Granted To: {permission.GrantedTo.User.DisplayName} ({permission.GrantedTo.User.Id})");
                    }

                    if (permission.GrantedToIdentities != null)
                    {
                        foreach (var identity in permission.GrantedToIdentities)
                        {
                            Console.WriteLine($"  Granted To: {identity.User?.DisplayName} ({identity.User?.Id})");
                        }
                    }

                    Console.WriteLine(new string('-', 80));
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error listing permissions: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Deletes a permission
        /// </summary>
        /// <param name="siteId">The site ID</param>
        /// <param name="driveId">The drive ID</param>
        /// <param name="itemId">The item ID</param>
        /// <param name="permissionId">The permission ID to delete</param>
        public async Task DeletePermissionAsync(string siteId, string driveId, string itemId, string permissionId)
        {
            try
            {
                Console.WriteLine($"Deleting permission: {permissionId}");

                await _graphClient
                    .Drives[driveId]
                    .Items[itemId]
                    .Permissions[permissionId]
                    .DeleteAsync();

                Console.WriteLine("Permission deleted successfully");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error deleting permission: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Lists the permissions of a site (read-only).
        /// </summary>
        public async Task ListSitePermissionsAsync(string siteId)
        {
            try
            {
                Console.WriteLine($"Listing permissions of site: {siteId}");
                var permissions = await _graphClient.Sites[siteId].Permissions.GetAsync();

                Console.WriteLine($"Found {permissions?.Value?.Count ?? 0} site permissions:");
                Console.WriteLine(new string('-', 80));
                foreach (var permission in permissions?.Value ?? new System.Collections.Generic.List<Permission>())
                {
                    Console.WriteLine($"Permission ID: {permission.Id}");
                    Console.WriteLine($"  Roles: {string.Join(", ", permission.Roles ?? new System.Collections.Generic.List<string>())}");
                    if (permission.GrantedToIdentities != null)
                    {
                        foreach (var identity in permission.GrantedToIdentities)
                        {
                            Console.WriteLine($"  Granted To: {identity.User?.DisplayName}");
                        }
                    }
                    Console.WriteLine(new string('-', 80));
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error listing site permissions: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Entry point: builds the client with the common module and runs a read-only
        /// demo listing the permissions of the book-test test site.
        /// </summary>
        public static async Task Main(string[] args)
        {
            try
            {
                var graphClient = SharePointGraphClientFactory.CreateFromSecret();
                var permOps = new PermissionOperations(graphClient);

                var hostname = Environment.GetEnvironmentVariable("SHAREPOINT_HOSTNAME") ?? "olddogsoft1.sharepoint.com";
                var sitePath = Environment.GetEnvironmentVariable("SHAREPOINT_SITE_PATH") ?? "book-test";

                // Resolve the site by path to get its ID.
                var site = await graphClient.Sites[$"{hostname}:/sites/{sitePath}"].GetAsync();
                if (site == null)
                {
                    Console.WriteLine("Test site not found.");
                    return;
                }
                var siteId = site.Id ?? throw new InvalidOperationException("El sitio no tiene ID.");

                // List permissions for an ITEM in the test site drive (within the
                // Sites.Selected grant). Listing /sites/{id}/permissions requires admin
                // permissions (Sites.Manage/FullControl), outside the least-privilege scope.
                var drives = await graphClient.Sites[siteId].Drives.GetAsync();
                var firstDrive = drives?.Value?.FirstOrDefault();
                if (firstDrive == null)
                {
                    Console.WriteLine("No se encontraron bibliotecas en el sitio de pruebas.");
                    return;
                }
                var driveId = firstDrive.Id ?? throw new InvalidOperationException("La biblioteca no tiene ID.");
                await permOps.ListPermissionsAsync(siteId, driveId, "root");

                Console.WriteLine("\nPermission operations completed successfully!");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error: {ex.Message}");
                Environment.Exit(1);
            }
        }
    }
}

File: chapter-05-permissions/java/01-permissions/PermissionOperations.java
// PermissionOperations.java
// Chapter 05: Permissions
//
// Ejemplo de operaciones de permisos y sharing en SharePoint con Microsoft Graph
// SDK para Java (v6). El GraphServiceClient se inyecta por constructor (DI) desde
// el módulo común: demostración de DRY + inyección de dependencias.

package com.sharepointexperience.chapter05;

import com.laexperiencia.sharepoint.common.GraphServiceClientFactory;
import com.microsoft.graph.drives.item.items.item.createlink.CreateLinkPostRequestBody;
import com.microsoft.graph.drives.item.items.item.invite.InvitePostRequestBody;
import com.microsoft.graph.models.DriveRecipient;
import com.microsoft.graph.models.Permission;
import com.microsoft.graph.models.PermissionCollectionResponse;
import com.microsoft.graph.models.SharingLink;
import com.microsoft.graph.models.Site;
import com.microsoft.graph.serviceclient.GraphServiceClient;

import java.util.ArrayList;
import java.util.List;

/**
 * Ejemplo que demuestra gestión de permisos y enlaces de uso compartido en
 * SharePoint: creación de enlaces anónimos/organización, concesión de acceso a
 * usuarios, listado y borrado de permisos a nivel de item y de sitio.
 */
public class PermissionOperations {

    private final GraphServiceClient graphClient;

    /** Crea una instancia con un cliente de Graph inyectado (DI). */
    public PermissionOperations(GraphServiceClient graphClient) {
        if (graphClient == null) {
            throw new IllegalArgumentException("graphClient is required");
        }
        this.graphClient = graphClient;
    }

    /**
     * Crea un enlace de uso compartido anónimo (cualquiera con el enlace) para un
     * archivo o carpeta.
     */
    public Permission createAnonymousLink(String driveId, String itemId, String linkType) {
        String type = linkType == null || linkType.isBlank() ? "view" : linkType;
        System.out.println("Creando enlace anónimo (" + type + ") para el item: " + itemId);

        CreateLinkPostRequestBody body = new CreateLinkPostRequestBody();
        body.setType(type);
        body.setScope("anonymous");

        Permission permission = graphClient
                .drives().byDriveId(driveId)
                .items().byDriveItemId(itemId)
                .createLink().post(body);

        printLink(permission);
        return permission;
    }

    /**
     * Crea un enlace de uso compartido para toda la organización.
     */
    public Permission createOrganizationLink(String driveId, String itemId, String linkType) {
        String type = linkType == null || linkType.isBlank() ? "view" : linkType;
        System.out.println("Creando enlace de organización (" + type + ") para el item: " + itemId);

        CreateLinkPostRequestBody body = new CreateLinkPostRequestBody();
        body.setType(type);
        body.setScope("organization");

        Permission permission = graphClient
                .drives().byDriveId(driveId)
                .items().byDriveItemId(itemId)
                .createLink().post(body);

        printLink(permission);
        return permission;
    }

    /**
     * Concede acceso a un usuario concreto sobre un item.
     */
    public Permission grantAccessToUser(String driveId, String itemId, String userEmail, String role) {
        String effectiveRole = role == null || role.isBlank() ? "write" : role;
        System.out.println("Concediendo acceso '" + effectiveRole + "' a " + userEmail + " sobre el item: " + itemId);

        DriveRecipient recipient = new DriveRecipient();
        recipient.setEmail(userEmail);

        List<DriveRecipient> recipients = new ArrayList<>();
        recipients.add(recipient);

        List<String> roles = new ArrayList<>();
        roles.add(effectiveRole);

        InvitePostRequestBody body = new InvitePostRequestBody();
        body.setRecipients(recipients);
        body.setRoles(roles);
        body.setSendInvitation(true);
        body.setMessage("You have been granted access to this document.");

        var permissions = graphClient
                .drives().byDriveId(driveId)
                .items().byDriveItemId(itemId)
                .invite().post(body);

        Permission granted = (permissions != null && permissions.getValue() != null && !permissions.getValue().isEmpty())
                ? permissions.getValue().get(0)
                : null;

        if (granted != null) {
            System.out.println("Acceso concedido:");
            System.out.println("  Permission ID: " + granted.getId());
            System.out.println("  Roles: " + granted.getRoles());
        }
        return granted;
    }

    /**
     * Lista los permisos de un item (drive).
     */
    public void listPermissions(String driveId, String itemId) {
        System.out.println("Listando permisos del item: " + itemId);

        PermissionCollectionResponse response = graphClient
                .drives().byDriveId(driveId)
                .items().byDriveItemId(itemId)
                .permissions().get();

        List<Permission> permissions = response != null && response.getValue() != null
                ? response.getValue()
                : new ArrayList<>();

        System.out.println("Encontrados " + permissions.size() + " permisos:");
        System.out.println("-".repeat(80));
        for (Permission p : permissions) {
            printPermission(p);
            System.out.println("-".repeat(80));
        }
    }

    /**
     * Lista los permisos a nivel de sitio (solo lectura).
     */
    public void listSitePermissions(String siteId) {
        System.out.println("Listando permisos del sitio: " + siteId);

        PermissionCollectionResponse response = graphClient
                .sites().bySiteId(siteId)
                .permissions().get();

        List<Permission> permissions = response != null && response.getValue() != null
                ? response.getValue()
                : new ArrayList<>();

        System.out.println("Encontrados " + permissions.size() + " permisos de sitio:");
        System.out.println("-".repeat(80));
        for (Permission p : permissions) {
            printPermission(p);
            System.out.println("-".repeat(80));
        }
    }

    /**
     * Elimina un permiso concreto de un item.
     */
    public void deletePermission(String driveId, String itemId, String permissionId) {
        System.out.println("Eliminando permiso: " + permissionId);
        graphClient
                .drives().byDriveId(driveId)
                .items().byDriveItemId(itemId)
                .permissions().byPermissionId(permissionId)
                .delete();
        System.out.println("Permiso eliminado correctamente");
    }

    // ============== Auxiliares ==============

    private void printLink(Permission permission) {
        if (permission == null) {
            System.out.println("  (sin respuesta)");
            return;
        }
        SharingLink link = permission.getLink();
        if (link != null) {
            System.out.println("  Enlace creado:");
            System.out.println("    Web URL: " + link.getWebUrl());
            System.out.println("    Type: " + link.getType());
            System.out.println("    Scope: " + link.getScope());
        }
    }

    private void printPermission(Permission p) {
        System.out.println("Permission ID: " + p.getId());
        System.out.println("  Roles: " + p.getRoles());
        if (p.getLink() != null) {
            System.out.println("  Link Type: " + p.getLink().getType());
            System.out.println("  Link Scope: " + p.getLink().getScope());
            System.out.println("  Web URL: " + p.getLink().getWebUrl());
        }
        if (p.getGrantedToIdentitiesV2() != null) {
            for (var identity : p.getGrantedToIdentitiesV2()) {
                System.out.println("  Granted To (v2): "
                        + (identity.getApplication() != null ? identity.getApplication().getDisplayName()
                            : (identity.getUser() != null ? identity.getUser().getDisplayName() : "N/A")));
            }
        } else if (p.getGrantedTo() != null && p.getGrantedTo().getUser() != null) {
            System.out.println("  Granted To: " + p.getGrantedTo().getUser().getDisplayName());
        }
    }

    /**
     * Entry point: construye el cliente con el módulo común y ejecuta una demo de
     * solo lectura listando los permisos del sitio de pruebas book-test (resuelto
     * por path, sobreescribible con SHAREPOINT_HOSTNAME / SHAREPOINT_SITE_PATH).
     */
    public static void main(String[] args) {
        try {
            // create() auto-detecta certificado (si CERTIFICATE_PATH está presente) o client secret.
            GraphServiceClient client = GraphServiceClientFactory.create();
            PermissionOperations permOps = new PermissionOperations(client);

            String hostname = envOr("SHAREPOINT_HOSTNAME", "olddogsoft1.sharepoint.com");
            String sitePath = envOr("SHAREPOINT_SITE_PATH", "book-test");

            // Resolver el sitio por path para obtener su ID.
            Site site = client.sites().bySiteId(hostname + ":/sites/" + sitePath).get();
            String siteId = site.getId();
            System.out.println("Sitio: " + site.getDisplayName() + " (" + siteId + ")\n");

            // Listar los drives del sitio y tomar el primero; listar permisos del item root.
            // (Listar /sites/{id}/permissions requiere permisos de admin fuera de Sites.Selected;
            //  los permisos de un item del drive sí están cubiertos por el grant Sites.Selected.)
            var drives = client.sites().bySiteId(siteId).drives().get();
            var firstDrive = (drives != null && drives.getValue() != null && !drives.getValue().isEmpty())
                    ? drives.getValue().get(0) : null;
            if (firstDrive == null) {
                System.out.println("No se encontraron bibliotecas en el sitio de pruebas.");
                return;
            }
            System.out.println("Biblioteca: " + firstDrive.getName() + " (" + firstDrive.getId() + ")\n");
            try {
                permOps.listPermissions(firstDrive.getId(), "root");
            } catch (Exception ex) {
                // El SDK de Graph para Java tiene un quirk de parseo de fechas (OffsetDateTime
                // sin zona) en algunas respuestas; lo reportamos sin abortar el ejemplo.
                System.err.println("⚠️ No se pudieron listar los permisos: " + ex.getMessage());
            }

            System.out.println("\nPermission operations completed successfully!");
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }

    private static String envOr(String name, String defaultValue) {
        String v = System.getenv(name);
        return (v == null || v.isBlank()) ? defaultValue : v;
    }
}